### ISSUE TITLE
Sidecar stopped submitting stats to StackDriver after prometheus eviction on GKE

### ISSUE THREAD

**Issue by lkoba** [Stackdriver/stackdriver-prometheus-sidecar#83]
We had stackdriver-prometheus-sidecar working on our GKE cluster without issues for a few days until the prometheus pod where the server and this sidecar resided was evicted. The pod was relocated automatically by the scheduler and Prometheus continued collecting data without issues. However, the sidecar stopped submitting data to StackDriver. We can manually query the data from Prometheus without issues, there's no discernible missing data.

Versions:
prom/prometheus:v2.6.1
gcr.io/stackdriver-prometheus/stackdriver-prometheus-sidecar:release-0.3.1

Prometheus Logs:
```
level=info ts=2019-01-26T15:28:19.524238548Z caller=main.go:243 msg="Starting Prometheus" version="(version=2.6.1, branch=HEAD, revision=b639fe140c1f71b2cbad3fc322b17efe60839e7e)"
level=info ts=2019-01-26T15:28:19.524313035Z caller=main.go:244 build_context="(go=go1.11.4, user=root@4c0e286fe2b3, date=20190115-19:12:04)"
level=info ts=2019-01-26T15:28:19.524350038Z caller=main.go:245 host_details="(Linux 4.15.0-1019-gcp #20-Ubuntu SMP Wed Aug 29 09:24:47 UTC 2018 x86_64 prometheus-server-c64fd76c-msmv5 (none))"
level=info ts=2019-01-26T15:28:19.524380622Z caller=main.go:246 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2019-01-26T15:28:19.524485977Z caller=main.go:247 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2019-01-26T15:28:19.525637533Z caller=main.go:561 msg="Starting TSDB ..."
level=info ts=2019-01-26T15:28:19.527003933Z caller=web.go:429 component=web msg="Start listening for connections" address=0.0.0.0:9090
level=info ts=2019-01-26T15:28:19.574606943Z caller=repair.go:48 component=tsdb msg="found healthy block" mint=1548244800000 maxt=1548266400000 ulid=01D1Y8722N7PQF05HFZWCSFMK2
level=info ts=2019-01-26T15:28:19.599054445Z caller=repair.go:48 component=tsdb msg="found healthy block" mint=1548266400000 maxt=1548331200000 ulid=01D2060M2RPVQJBJ53YRAQNRV6
level=info ts=2019-01-26T15:28:19.601827225Z caller=repair.go:48 component=tsdb msg="found healthy block" mint=1548331200000 maxt=1548396000000 ulid=01D223SQDT2MRJR8E8XMKXJ3BE
level=info ts=2019-01-26T15:28:19.621762297Z caller=repair.go:48 component=tsdb msg="found healthy block" mint=1548396000000 maxt=1548460800000 ulid=01D241K5Z5ZM6GHAJ6QFZX8PJY
level=info ts=2019-01-26T15:28:19.622993701Z caller=repair.go:48 component=tsdb msg="found healthy block" mint=1548460800000 maxt=1548482400000 ulid=01D24P696MFWW21CJZJGNEC14Q
level=info ts=2019-01-26T15:28:19.635275773Z caller=repair.go:48 component=tsdb msg="found healthy block" mint=1548504000000 maxt=1548511200000 ulid=01D25ASCGQ3N7HAHP75Q15SRVX
level=info ts=2019-01-26T15:28:19.636098417Z caller=repair.go:48 component=tsdb msg="found healthy block" mint=1548482400000 maxt=1548504000000 ulid=01D25ASDY6Y1S5K0AJZSCYB6GR
level=info ts=2019-01-26T15:28:39.425314438Z caller=main.go:571 msg="TSDB started"
level=info ts=2019-01-26T15:28:39.425436903Z caller=main.go:631 msg="Loading configuration file" filename=/etc/config/prometheus.yml
level=info ts=2019-01-26T15:28:39.683670989Z caller=kubernetes.go:201 component="discovery manager scrape" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2019-01-26T15:28:39.734034519Z caller=kubernetes.go:201 component="discovery manager notify" discovery=k8s msg="Using pod service account via in-cluster config"
level=info ts=2019-01-26T15:28:39.735718748Z caller=main.go:657 msg="Completed loading of configuration file" filename=/etc/config/prometheus.yml
level=info ts=2019-01-26T15:28:39.735774879Z caller=main.go:530 msg="Server is ready to receive web requests."
```

Sidecar Logs
```
level=info ts=2019-01-26T15:28:18.633697598Z caller=main.go:256 msg="Starting Stackdriver Prometheus sidecar" version="(version=0.3.1, branch=release-0.3.1, revision=12aa811802effd7a93b6b7d83d72bf381f1179b9)"
level=info ts=2019-01-26T15:28:18.633825024Z caller=main.go:257 build_context="(go=go1.11, user=jkohen@nadax.c.googlers.com, date=20190103-20:48:32)"
level=info ts=2019-01-26T15:28:18.633857706Z caller=main.go:258 host_details="(Linux 4.15.0-1019-gcp #20-Ubuntu SMP Wed Aug 29 09:24:47 UTC 2018 x86_64 prometheus-server-c64fd76c-msmv5 (none))"
level=info ts=2019-01-26T15:28:18.633878294Z caller=main.go:259 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2019-01-26T15:28:18.646053105Z caller=main.go:463 msg="Web server started"
level=info ts=2019-01-26T15:28:18.646116703Z caller=main.go:444 msg="Stackdriver client started"
level=warn ts=2019-01-26T15:28:21.653132839Z caller=main.go:525 msg="Prometheus not ready" status="503 Service Unavailable"
level=warn ts=2019-01-26T15:28:24.652898547Z caller=main.go:525 msg="Prometheus not ready" status="503 Service Unavailable"
level=warn ts=2019-01-26T15:28:27.651712449Z caller=main.go:525 msg="Prometheus not ready" status="503 Service Unavailable"
level=warn ts=2019-01-26T15:28:30.652042511Z caller=main.go:525 msg="Prometheus not ready" status="503 Service Unavailable"
level=warn ts=2019-01-26T15:28:33.652120435Z caller=main.go:525 msg="Prometheus not ready" status="503 Service Unavailable"
level=warn ts=2019-01-26T15:28:36.652012708Z caller=main.go:525 msg="Prometheus not ready" status="503 Service Unavailable"
level=warn ts=2019-01-26T15:28:39.652032741Z caller=main.go:525 msg="Prometheus not ready" status="503 Service Unavailable"
level=info ts=2019-01-26T15:29:42.652116635Z caller=manager.go:150 component="Prometheus reader" msg="Starting Prometheus reader..."
```

We can see through docker stats that the CPU activity on prometheus and the sidecar is around 800m each. If we remove the sidecar, Prometheus drops to 5-10m.

I'm guessing that this may be related to the repair.go messages on Prometheus' logs so we let the sidecar work. However, after a few hours there's still no progress or logs. Our data dir is around 600mb, it shouldn't take that long so it seems that it's stuck.

We have added   --log.level=debug to the sidecar but there's no difference except during the shutdown phase that shows a "Stopped resharding":
```
level=warn ts=2019-01-26T16:39:23.977263853Z caller=main.go:392 msg="Received SIGTERM, exiting gracefully..."
level=info ts=2019-01-26T16:39:23.977387876Z caller=main.go:432 msg="Stopping Prometheus reader..."
level=info ts=2019-01-26T16:39:23.97741396Z caller=queue_manager.go:233 component=queue_manager msg="Stopping remote storage..."
level=debug ts=2019-01-26T16:39:23.977513514Z caller=queue_manager.go:441 component=queue_manager msg="Stopped resharding"
level=info ts=2019-01-26T16:39:23.977599116Z caller=queue_manager.go:241 component=queue_manager msg="Remote storage stopped."
```

**Comment by lkoba** [comment]
I added the sidecar as a target but those metrics do not show up.

The only metric I can see with the prometheus_sidecar_ prefix is prometheus_sidecar_dropped_series, I'm attaching the graph below and a manual scrape of the metrics endpoint of the sidecar.

![screen shot 2019-01-27 at 12 49 00](https://user-images.githubusercontent.com/168509/51803313-73526680-2232-11e9-84bd-6104a3cdaeee.png)

Sidecar Metrics:
[sd_metrics.txt](https://github.com/Stackdriver/stackdriver-prometheus-sidecar/files/2800207/sd_metrics.txt)

**Comment by lkoba** [comment]
I just realized that this last information that I posted may be useless. Before giving up I deleted the stackdriver_sidecar.json and let is work overnight to see if it made a difference.

**Comment by lkoba** [comment]
No. I posted the information requested but didn't research much more, I ended up using prometheus-to-sd as a workaround to get the information I needed on StackDriver. Thanks for your help.

**Comment by fabxc** [comment]
Hi @lkoba, thanks for the graph and the metric dump.

This indicates that the metrics the sidecar could not find a matching target via Prometheus's API for the metrics it read from disk. It also clearly does a lot of requests to Prometheus target API endpoint.

Could you also provide is with the curl output from `/api/v1/targets`?
Your full configuration file for Prometheus would also be helpful.

**Comment by lkoba** [comment]
Here's the information you requested:

[targets.json.txt](https://github.com/Stackdriver/stackdriver-prometheus-sidecar/files/2829647/targets.json.txt)

[prometheus.yml.txt](https://github.com/Stackdriver/stackdriver-prometheus-sidecar/files/2829653/prometheus.yml.txt)

Please let me know if you need anything else.

**Comment by fabxc** [comment]
Hi @lkoba,

we've likely identified the issue and will push a fix soon. Once it's in, you might want to give this another shot.

**Comment by lkoba** [comment]
Thanks! Just let me know when, and I'll test it.

**Comment by StevenYCChou** [comment]
Hi @lkoba,

We've identified an issue that negative cache in targets doesn't work as we expected, and it may make sidecar keep querying targets API and block the traffic. We fixed this issue with #91.

We've released a new release [v0.4.0](https://github.com/Stackdriver/stackdriver-prometheus-sidecar/releases/tag/v0.4.0) including #91, and you may want to try it out.

Thanks again for reporting the issue!

**Comment by StevenYCChou** [comment]
Hi @lkoba, 

Just want to follow up on this thread. If you ever test the new image, let us know whether you encounter any issue. My team will be happy to hear your feedback!
